torchdynamo/eval_frame.py

```diff
@@ -279,16 +279,28 @@ def dynamo_normalization_capturing_compiler(
         out_sig.parameters[k].annotation for k in list(out_sig.parameters)
     ]
 
-    # TODO(voz): Add support for flatenning and unflattening via PyTree
-    assert len(in_sig.parameters) == len(
-        out_sig.parameters
-    ), "Exported callable signature must be composed only of torch.Tensors"
-    for idx in range(len(out_sig.parameters)):
-        sig_type = signature_types[idx]
-        in_type = input_types[idx]
-        assert (
-            sig_type == in_type
-        ), "Export produced a graph with mismatched type signature {sig_type} vs expected {in_type} for arg {idx}"
+    def assert_equal_params(in_types):
+        for idx in range(len(out_sig.parameters)):
+            sig_type = signature_types[idx]
+            in_type = in_types[idx]
+            assert (
+                sig_type == in_type
+            ), "Export produced a graph with mismatched type signature {sig_type} vs expected {in_type} for arg {idx}"
+
+    if len(in_sig.parameters) == len(out_sig.parameters):
+        assert_equal_params(input_types)
+    else:
+        from functorch._src.aot_autograd import pytree
+
+        flat_input_types = []
+        flat_args, _ = pytree.tree_flatten(args, kwargs)
+        for arg in flat_args:
+            flat_input_types.append(arg.__class__)
+
+        assert len(flat_input_types) == len(
+            out_sig.parameters
+        ), "Flattened inputs length must match out signature parameter lengths."
+        assert_equal_params(flat_input_types)
 
     return (graph, out_guards)
```

Comment:
How would the user call the flattened graph i we don't return the pytree?

Reply:
Good question. Where I integrate it now, we call pytree on the caller side, and pass in the flattened args again.